### PR TITLE
Adapt man-page making for OpenSSL master / 3.0

### PR DESCRIPTION
**Pull request by levitte**
We use OpenSSL's rendering instead of our own, and just lightly strip
the result to fit in our page layout.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -79,48 +82,87 @@ rebuild: all
...
+$(eval $(call makemanindexes,$(1),$(2)))
+endef
 
 # Now that we have the generating macros in place, let's use them!
 #
 # Start off with creating the 'manpages-master' target, taking the
 # source from $(CHECKOUTS)/openssl/doc
-$(eval $(call makemanpages,openssl,master))
+$(eval $(call makemanuals3,openssl,master))
+#$(foreach S,$(MANSERIES3),$(eval $(call makemanuals3,openssl-$(S),$(S))))
```
Comment (levitte) [inline Makefile:151]: This is to be uncommented when we have a 3.0 branch

Comment (romen) [reply Makefile:151]: Should this comment be included in the Makefile itself?

**Comment by romen** [comment]
I believe this would be the fix for https://github.com/openssl/tools/issues/66

Thanks @levitte for looking into this!

**Comment by romen** [comment]
Ping @mattcaswell for a review.

I believe this should go in as soon as possible, so that the manpages documenting master/alpha1 are not crippled now that the release is out!

**Comment by romen** [comment]
Ping @openssl/omc

**Comment by romen** [comment]
@levitte
I am trying to use the Makefile to generate my own local mirror of the docs as they appear on openssl.org

I notice `bin/strip-man-html` is not contained in the repo, is this intentional?

**Comment by romen** [comment]
`bin/mk-apropos` is looking for metadata that is included in the `man1.1.1` pages but is not in `manmaster`:

https://github.com/openssl/web/blob/36975fbb439f2e93a907941519c9c66787e6a3d2/bin/mk-apropos#L8-L10

Here is a diff of an html between `man1.1.1` and `manmaster`:

```diff
--- docs/man1.1.1/man3/EVP_PKEY_cmp.html	2020-05-11 16:40:47.860246592 +0200
+++ docs/manmaster/man3/EVP_PKEY_cmp.html	2020-05-12 12:44:55.195544478 +0200
@@ -1,47 +1,49 @@
-<!DOCTYPE html>
-<html lang="en">
-<!-- OSSL: original subdir: man3 -->
-<!-- OSSL: subdir: man3 -->
-<!-- OSSL: section: 3 -->
-<!-- OSSL: description: public key parameter and comparison functions  -->
-<!--#include virtual="/inc/head.shtml" -->
-<body>
-<!--#include virtual="/inc/banner.shtml" -->
-  <div id="main">
-    <div id="content">
-      <div class="blog-index">
-        <article>
-          <header><h2>EVP_PKEY_cmp</h2></header>
-          <div class="entry-content">
-            <p>
-<!--#include virtual="EVP_PKEY_cmp.inc" -->
-            </p>
-          </div>
-          <footer>
-            You are here: <a href="/">Home</a>
-            : <a href="/docs">Docs</a>
-            : <a href="/docs/manpages.html">Manpages</a>
-            : <a href="/docs/man1.1.1/">1.1.1</a>
-            : <a href="/docs/man1.1.1/man3">man3</a>
-            : <a href="/docs/man1.1.1/man3/EVP_PKEY_cmp.html">EVP_PKEY_cmp</a>
-            <br/><a href="/sitemap.txt">Sitemap</a>
-          </footer>
-        </article>
-      </div>
-      <aside class="sidebar">
-        <section>
-          <h1><a href="/docs/man1.1.1/">1.1.1 manpages</a></h1>
-          <ul>
-            <li><a href="../man1">Commands</a></li>
-            <li><a href="../man3">Libraries</a></li>
-            <li><a href="../man5">File Formats</a></li>
-            <li><a href="../man7">Overviews</a></li>
-          </ul>
-        </section>
-<!--#include virtual="EVP_PKEY_cmp.cross" -->
-      </aside>
-    </div>
-  </div>
-<!--#include virtual="/inc/footer.shtml" -->
-</body>
-</html>
+<h3 id="NAME">NAME</h3>
+
+<p>EVP_PKEY_copy_parameters, EVP_PKEY_missing_parameters, EVP_PKEY_cmp_parameters, EVP_PKEY_cmp - public key parameter and comparison functions</p>
+
+<h3 id="SYNOPSIS">SYNOPSIS</h3>
+
+<pre><code> #include &lt;openssl/evp.h&gt;
+
+ int EVP_PKEY_missing_parameters(const EVP_PKEY *pkey);
+ int EVP_PKEY_copy_parameters(EVP_PKEY *to, const EVP_PKEY *from);
+
+ int EVP_PKEY_cmp_parameters(const EVP_PKEY *a, const EVP_PKEY *b);
+ int EVP_PKEY_cmp(const EVP_PKEY *a, const EVP_PKEY *b);</code></pre>
+
+<h3 id="DESCRIPTION">DESCRIPTION</h3>
+
+<p>The function EVP_PKEY_missing_parameters() returns 1 if the public key parameters of <b>pkey</b> are missing and 0 if they are present or the algorithm doesn&#39;t use parameters.</p>
+
+<p>The function EVP_PKEY_copy_parameters() copies the parameters from key <b>from</b> to key <b>to</b>. An error is returned if the parameters are missing in <b>from</b> or present in both <b>from</b> and <b>to</b> and mismatch. If the parameters in <b>from</b> and <b>to</b> are both present and match this function has no effect.</p>
+
+<p>The function EVP_PKEY_cmp_parameters() compares the parameters of keys <b>a</b> and <b>b</b>.</p>
+
+<p>The function EVP_PKEY_cmp() compares the public key components and parameters (if present) of keys <b>a</b> and <b>b</b>.</p>
+
+<h3 id="NOTES">NOTES</h3>
+
+<p>The main purpose of the functions EVP_PKEY_missing_parameters() and EVP_PKEY_copy_parameters() is to handle public keys in certificates where the parameters are sometimes omitted from a public key if they are inherited from the CA that signed it.</p>
+
+<p>Since OpenSSL private keys contain public key components too the function EVP_PKEY_cmp() can also be used to determine if a private key matches a public key.</p>
+
+<h3 id="RETURN-VALUES">RETURN VALUES</h3>
+
+<p>The function EVP_PKEY_missing_parameters() returns 1 if the public key parameters of <b>pkey</b> are missing and 0 if they are present or the algorithm doesn&#39;t use parameters.</p>
+
+<p>These functions EVP_PKEY_copy_parameters() returns 1 for success and 0 for failure.</p>
+
+<p>The function EVP_PKEY_cmp_parameters() and EVP_PKEY_cmp() return 1 if the keys match, 0 if they don&#39;t match, -1 if the key types are different and -2 if the operation is not supported.</p>
+
+<h3 id="SEE-ALSO">SEE ALSO</h3>
+
+<p><a href="../man3/EVP_PKEY_CTX_new.html>EVP_PKEY_CTX_new(3)</a>, <a href="../man3/EVP_PKEY_keygen.html>EVP_PKEY_keygen(3)</a></p>
+
+<h3 id="COPYRIGHT">COPYRIGHT</h3>
+
+<p>Copyright 2006-2016 The OpenSSL Project Authors. All Rights Reserved.</p>
+
+<p>Licensed under the Apache License 2.0 (the &quot;License&quot;). You may not use this file except in compliance with the License. You can obtain a copy in the file LICENSE in the source distribution or at <a href="https://www.openssl.org/source/license.html">https://www.openssl.org/source/license.html</a>.</p>
+
+

```

**Comment by romen** [comment]
I believe the output of `bin/strip-man-html` should go into a `.inc` file, not the `.html` file.

We still need to generate the template `.html` page with something similar to 
https://github.com/openssl/web/blob/36975fbb439f2e93a907941519c9c66787e6a3d2/bin/mk-manpages#L93-L145

I have  Makefile patch to do the first part, but I have no idea about how to run only a subroutine from a perl script to do the second part!

@levitte do you have a fix?

**Review by mattcaswell — COMMENTED** [review]
Looks good except for the comments noted by @romen

**Comment by levitte** [comment]
Good point.  As for generating the .html, I need to create another simpler script that does only that

**Comment by levitte** [comment]
It will take me a bit, hopefully done some time today, maybe this evening

**Comment by romen** [comment]
Ping @levitte

**Comment by romen** [comment]
Ping

**Comment by levitte** [comment]
I had to do a bit of rework to get it all to fit and still be readable.

I believe this is ready now.  See https://beta.openssl.org/ for the live demo, it's currently built from this branch.

The page contents look horrific as soon as there's a link, but that should be fixed as soon as openssl/openssl#11969 comes through.

File: bin/all-html-man-names [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#! /usr/bin/env perl
+
+
+local $/;                       # Slurp the whole file
+my $contents = <STDIN>;
+
+$contents =~ m@^<h3 id="NAME">NAME</h3>@m;
+$contents = $';       # Everything before and including NAME is remove
```
Comment (mattcaswell) [inline bin/all-html-man-names]: s/remove/removed/

**Review by mattcaswell — APPROVED** [review]
LGTM. Approved assuming the minor nit is fixed.

File: bin/all-html-man-names (hunk 2 above)
Comment (levitte) [reply bin/all-html-man-names]: Done

**Comment by levitte** [comment]
openssl/openssl#11969 is in, and I just tried a re-run on beta.openssl.org, and things look fine now.  Time to merge this.

**Comment by levitte** [comment]
Merged

ea973d250e311c51c91217c2e6edf93370be0e43 Adapt man-page making for OpenSSL master / 3.0

**Pull request closed without merging**